dask/base.py

```diff
@@ -132,14 +132,42 @@ def normalize_function(func):
 normalize_token.register(object,
         lambda a: normalize_function(a) if callable(a) else a)
 normalize_token.register(dict, lambda a: tuple(sorted(a.items())))
+
 with ignoring(ImportError):
     import pandas as pd
-    normalize_token.register(pd.DataFrame,
-            lambda a: (id(a), len(a), list(a.columns)))
-    normalize_token.register(pd.Series, lambda a: (id(a), len(a), a.name))
+
+    @partial(normalize_token.register, pd.Index)
+    def normalize_index(ind):
+        return tokenize(ind.name, ind.values)
+
+    @partial(normalize_token.register, pd.Categorical)
+    def normalize_categorical(cat):
+        return tokenize(cat.codes, cat.categories, cat.ordered)
+
+    @partial(normalize_token.register, pd.Series)
+    def normalize_series(s):
+        return tokenize(s.name, s._data.blocks[0].values, s.dtype, s.index)
```

Comment:
Ah, although I suppose that this doesn't count extra calls to all of the tokenize machinery.

```python
In [1]: from dask.base import tokenize

In [2]: %time _ = tokenize(['1y9hfshfs90f8hfs0shf', 'hf90s8hfsh90h098hf3'])
CPU times: user 17 µs, sys: 2 µs, total: 19 µs
Wall time: 22.2 µs

In [3]: %timeit tokenize(['1y9hfshfs90f8hfs0shf', 'hf90s8hfsh90h098hf3'])
The slowest run took 7.86 times longer than the fastest. This could mean that an intermediate result is being cached 
100000 loops, best of 3: 2.79 µs per loop
```